staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go

```diff
@@ -172,34 +158,89 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 	if p.ResourceVersion != nil {
 		patch, err = addResourceVersion(patch, *p.ResourceVersion)
 		if err != nil {
-			return nil, nil, cmdutil.AddSourceToErr("Failed to insert resourceVersion in patch", source, err)
+			return nil, nil, errors.Wrap(err, "Failed to insert resourceVersion in patch")
 		}
 	}
 
 	patchedObj, err := p.Helper.Patch(namespace, name, patchType, patch, nil)
 	return patch, patchedObj, err
 }
 
+// buildMergePatch builds patch according to the JSONMergePatch which is used for
+// custom resource definitions.
+func (p *Patcher) buildMergePatch(original, modified, current []byte) ([]byte, error) {
+	preconditions := []mergepatch.PreconditionFunc{mergepatch.RequireKeyUnchanged("apiVersion"),
+		mergepatch.RequireKeyUnchanged("kind"), mergepatch.RequireMetadataKeyUnchanged("name")}
+	patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(original, modified, current, preconditions...)
+	if err != nil {
+		if mergepatch.IsPreconditionFailed(err) {
+			return nil, fmt.Errorf("%s", "At least one of apiVersion, kind and name was changed")
+		}
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// buildStrategicMergeFromOpenAPI builds patch from OpenAPI if it is enabled.
+// This is used for core types which is published in openapi.
+func (p *Patcher) buildStrategicMergeFromOpenAPI(gvk schema.GroupVersionKind, original, modified, current []byte) ([]byte, error) {
+	if s := p.OpenapiSchema.LookupResource(gvk); s != nil {
```

Comment:
I'd possibly remove this from the function. Returning `nil, nil` is not great.

Reply:
Since Go doesn't let you check if p.OpenapiSchema and assign at the same time, I'll suggest we add a function to the patcher like the following:
```golang
func (p *Patcher) findOpenAPIResource(gvk schema.GroupVersionKind) Resource {
    if p.OpenapiSchema == nil {
        return nil
    }
    return p.OpenapiSchema.LookupResource(gvk)
}
```

You can then use it above L138.

Reply:
@apelisse done, PTAL?